app/models/work_packages/scopes/relatable.rb

```diff
@@ -40,33 +40,114 @@ module Relatable
+      # Just for completeness of the example, both parent work packages are valid targets as well. But a predecessor of
+      # the predecessor's target would not be valid.
+      # It is also important to note, that existing ancestors are of no importance since this part of the structure will change.
+      # Creating a parent relationship is destructive since there can only ever be one.
+      #
+      # The result is a blocklist which will include work packages that can not be related to. The list is not complete
+      # as it will not include the work packages related by different relation types so those are added additionally. For the
+      # PARENT relationship, work packages directly related to any of the descendants are added as well. Since work packages
+      # related to the queried work packages only by a path containing both FOLLOWS and PRECEDES relationships are valid targets,
+      # these are filtered out as well. It would be possible to optimize the CTE here to not have them included
+      # in the first place.
+      #
+      # The CTE has the following columns:
+      # * id - the id of the work packages currently related. This is the result of the CTE.
+      # * from_(hierarchy/from_id/to_id) - booleans to prevent that the CTE returns back the path calculated in the previous
+      #                                    iteration.
+      # * origin - boolean to indicate whether the work package is the queried for work package or its ancestor/descendant
+      #            (only descendant for PARENT). Such a work package is never a valid target.
+      #            Additionally, it is used to limit proceeding in both directions for FOLLOWS/PRECEDES.
+      # * includes_(from_relation/to_relation) - booleans about the direction (from_id -> to_id or to_id -> from_id) of the path
+      #                                          (the relations followed).
+      #                                          This is relevant for a queried for PARENT relation. In that case, relations need
+      #                                          to be followed from the queried for work package (and its descendants) in both
+      #                                          directions. But only the direction taken from that origin needs to be followed
+      #                                          henceforth.
+      # * includes_hierarchy - boolean indicating that the last relation taken was a hierarchy relation. For a queried for
+      #                        PARENT relation, whenever that is the case, the work package is a valid relation target although
+      #                        it appears in the CTE.
       def relatable(work_package, relation_type)
         return all if work_package.new_record?
 
         scope = not_having_directed_relation(work_package, relation_type)
-                  .not_having_direct_relation(work_package)
```

Comment:
The names `not_having_directed_relation` and `not_having_direct_relation` are close. Are they in the same order as in the method documentation?

```
      # The implementation focuses on excluding candidates. It does so in two parts:
      #   * Excluding all work packages with which a direct relation already exist.
      #   * Excluding work packages that are related transitively.
```
It looks like they're not
And the order of usage is not the same as the order of definition below.

I like when order is the same everywhere: doc, invocation, definition. it makes it easier to understand. Can you swap `not_having_directed_relation` and `not_having_direct_relation` calls so that everything is in the same order?

And maybe renaming `not_having_directed_relation` as `not_having_transitive_relation` would better relate to the method description above.

Reply:
I restructured the code quite a bit in a7c3914b77a7b43c9a0bf5c19df2c10969889db1. The order in the code is now the same as it is in the documentation. The names changed significantly since the code now differentiates directly between PARENT relations and the rest.